File: app/src/main/java/org/tvheadend/tvhclient/features/information/WebViewFragment.java
package org.tvheadend.tvhclient.features.information;

import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.WebView;
import android.widget.ProgressBar;

import org.tvheadend.tvhclient.BuildConfig;
import org.tvheadend.tvhclient.R;
import org.tvheadend.tvhclient.features.shared.BaseFragment;
import org.tvheadend.tvhclient.features.shared.tasks.HtmlFileLoaderTask;
import org.tvheadend.tvhclient.utils.MiscUtils;

import java.util.regex.Pattern;

import androidx.annotation.NonNull;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;

public class WebViewFragment extends BaseFragment implements HtmlFileLoaderTask.Listener {

    @BindView(R.id.webview)
    protected WebView webView;
    @BindView(R.id.loading)
    protected ProgressBar progressBar;
    private HtmlFileLoaderTask htmlFileLoaderTask;
    private Unbinder unbinder;
    private String website;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.webview_fragment, container, false);
        unbinder = ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        forceSingleScreenLayout();

        if (savedInstanceState != null) {
            website = savedInstanceState.getString("website", "");
        } else {
            website = "";
            Bundle bundle = getArguments();
            if (bundle != null) {
                website = bundle.getString("website", "");
            }
        }

        switch (website) {
            case "information":
                toolbarInterface.setTitle(getString(R.string.pref_information));
                break;
            case "help_and_support":
                toolbarInterface.setTitle(getString(R.string.help_and_support));
                break;
            case "privacy_policy":
                toolbarInterface.setTitle(getString(R.string.pref_privacy_policy));
                break;
        }

        toolbarInterface.setSubtitle("");
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putString("website", website);
    }

    @Override
    public void onResume() {
        super.onResume();
        htmlFileLoaderTask = new HtmlFileLoaderTask(activity, website, "en", this);
        htmlFileLoaderTask.execute();
    }

    @Override
    public void onPause() {
        super.onPause();
        if (htmlFileLoaderTask != null) {
            htmlFileLoaderTask.cancel(true);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                activity.finish();
                return true;

            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public void onFileContentsLoaded(String content) {
        if (!TextUtils.isEmpty(content)) {

            if (content.contains("styles_light.css")) {
                if (MiscUtils.getThemeId(activity) == R.style.CustomTheme_Light) {
                    content = content.replace("styles_light.css", "html/styles_light.css");
                } else {
                    content = content.replace("styles_light.css", "html/styles_dark.css");
                }
            }
            if (content.contains("APP_VERSION")) {
                // Replace the placeholder in the html file with the real version
                String version = BuildConfig.VERSION_NAME + " (" + BuildConfig.BUILD_VERSION + ")";
                content = Pattern.compile("APP_VERSION").matcher(content).replaceAll(version);
            }

            webView.loadDataWithBaseURL("file:///android_asset/", content, "text/html", "utf-8", null);
            progressBar.setVisibility(View.GONE);
            webView.setVisibility(View.VISIBLE);
        }
    }
}

File: app/src/main/java/org/tvheadend/tvhclient/ui/features/information/WebViewFragment.kt
package org.tvheadend.tvhclient.ui.features.information

import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.view.*
import android.webkit.WebView
import android.widget.ProgressBar
import android.widget.TextView
import androidx.fragment.app.Fragment
import org.tvheadend.tvhclient.R
import org.tvheadend.tvhclient.ui.common.interfaces.FileContentsLoadedInterface
import org.tvheadend.tvhclient.ui.common.interfaces.LayoutControlInterface
import org.tvheadend.tvhclient.ui.common.interfaces.ToolbarInterface
import org.tvheadend.tvhclient.ui.features.settings.RemoveFragmentFromBackstackInterface
import org.tvheadend.tvhclient.util.extensions.gone
import org.tvheadend.tvhclient.util.extensions.visible
import org.tvheadend.tvhclient.util.extensions.visibleOrGone
import timber.log.Timber

open class WebViewFragment : Fragment(), FileContentsLoadedInterface {

    var website: String = ""
    private var webView: WebView? = null
    private var loadingView: ProgressBar? = null
    private var errorTextView: TextView? = null
    private lateinit var fileContentLoader: FileContentLoader
    lateinit var toolbarInterface: ToolbarInterface

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(if (Build.VERSION.SDK_INT in 21..25) R.layout.webview_fragment_for_lollipop else R.layout.webview_fragment, container, false)
        webView = view.findViewById(R.id.webview)
        loadingView = view.findViewById(R.id.loading_view)
        return view
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        if (activity is ToolbarInterface) {
            toolbarInterface = activity as ToolbarInterface
        }
        if (activity is LayoutControlInterface) {
            (activity as LayoutControlInterface).forceSingleScreenLayout()
        }

        fileContentLoader = FileContentLoader(requireContext(), "en", this)
        // Make the background transparent to remove flickering. This avoids seeing
        // the default theme background color before the stylesheets are loaded.
        webView?.setBackgroundColor(Color.argb(0, 0, 0, 0))
        webView?.gone()

        setHasOptionsMenu(true)
    }

    override fun onResume() {
        super.onResume()
        fileContentLoader.start(website)
    }

    override fun onPause() {
        super.onPause()
        fileContentLoader.stop()
    }

    override fun onPrepareOptionsMenu(menu: Menu) {
        super.onPrepareOptionsMenu(menu)
        menu.findItem(R.id.media_route_menu_item)?.isVisible = false
        menu.findItem(R.id.menu_search)?.isVisible = false
        menu.findItem(R.id.menu_reconnect_to_server)?.isVisible = false
        menu.findItem(R.id.menu_send_wake_on_lan_packet)?.isVisible = false
        menu.findItem(R.id.menu_privacy_policy)?.isVisible = false
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            android.R.id.home -> {
                if (activity is RemoveFragmentFromBackstackInterface) {
                    (activity as RemoveFragmentFromBackstackInterface).removeFragmentFromBackstack()
                }
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onFileContentsLoaded(content: String) {
        Timber.d("File contents loaded")
        loadingView?.gone()
        errorTextView?.visibleOrGone(content.isEmpty())

        if (content.isNotEmpty()) {
            webView?.loadDataWithBaseURL("file:///android_asset/", content, "text/html", "utf-8", null)
            webView?.visible()
        }
    }
}
